File: leetcode/easy/MergeTwoSortedLinkedLists.py
from typing import Optional


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Merges two sorted linked lists into one single sorted linked list.

        This solution uses an iterative approach with a dummy head node to build
        the new list. This avoids handling the edge case of an empty result list.

        Algorithm:
        1. Create a `dummy` node to serve as the starting point of the merged list.
        2. Create a `current` pointer, initialized to `dummy`, to track the tail of the merged list.
        3. Loop as long as both `list1` and `list2` have nodes.
        4. In each iteration, compare the values of the current nodes of `list1` and `list2`.
        5. Append the smaller node to `current.next` and advance the pointer of the list from which the node was taken.
        6. Move the `current` pointer to the newly added node.
        7. After the loop, one of the lists may still have remaining nodes. Append the
           non-empty list to the end of the merged list.
        8. Return `dummy.next`, which is the head of the merged list.

        Time Complexity: O(m + n), where m and n are the lengths of the two lists.
        Space Complexity: O(1), as we are just rearranging pointers, not creating new nodes.
        """
        # A dummy node acts as a placeholder to simplify the code,
        # avoiding edge cases for an empty merged list.
        dummy = ListNode()
        # `current` points to the last node in the merged list.
        current = dummy

        # Loop until one of the lists becomes empty.
        while list1 and list2:
            if list1.val < list2.val:
                # Append the smaller node from list1.
                current.next = list1
                # Move the list1 pointer forward.
                list1 = list1.next
            else:
                # Append the smaller or equal node from list2.
                current.next = list2
                # Move the list2 pointer forward.
                list2 = list2.next
            # Move the `current` pointer to the newly added node.
            current = current.next

        # Append the remaining part of the non-empty list.
        # This also correctly handles cases where one or both lists were initially empty.
        current.next = list1 or list2

        # The merged list starts at the node after the dummy head.
        return dummy.next

File: leetcode/easy/MergeTwoSortedLinkedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

/**
 * Solution for LeetCode problem "Merge Two Sorted Lists".
 * This class provides a method to merge two sorted singly-linked lists into a single sorted linked list.
 */
class Solution {

    /**
     * Merges two sorted singly-linked lists into a single sorted linked list.
     * The new list should be made by splicing together the nodes of the first two lists.
     *
     * This method uses an iterative approach with a dummy head node.
     * It compares the values of the current nodes from both lists and appends the smaller one
     * to the merged list, then advances the pointer of the list from which the node was taken.
     * After one list is exhausted, it appends the remaining nodes of the other list.
     *
     * @param list1 The head of the first sorted linked list.
     * @param list2 The head of the second sorted linked list.
     * @return The head of the merged sorted linked list.
     */
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        // Create a dummy node to simplify the handling of the head of the merged list.
        ListNode dummy = new ListNode();
        // 'current' pointer will be used to build the merged list.
        ListNode current = dummy;

        // Iterate while both lists have nodes.
        while (list1 != null && list2 != null) {
            // Compare the values of the current nodes from both lists.
            if (list1.val < list2.val) {
                // If list1's current node is smaller, append it to the merged list.
                current.next = list1;
                // Move to the next node in list1.
                list1 = list1.next;
            } else {
                // If list2's current node is smaller or equal, append it to the merged list.
                current.next = list2;
                // Move to the next node in list2.
                list2 = list2.next;
            }
            // Advance the 'current' pointer in the merged list.
            current = current.next;
        }

        // If one of the lists still has remaining elements, append them all to the merged list.
        // Since both input lists are sorted, the remaining elements are already in sorted order.
        if (list1 != null) {
            current.next = list1;
        } else {
            current.next = list2;
        }

        // The merged list starts from dummy.next (skipping the dummy node itself).
        return dummy.next;
    }
}
